src/Facebook/FileUpload/FacebookResumableUploader.php

```diff
@@ -0,0 +1,193 @@
+
+    /**
+     * @var string
+     */
+    protected $accessToken;
+
+    /**
+     * @var FacebookClient The Facebook client service.
+     */
+    protected $client;
+
+    /**
+     * @var string Graph version to use for this request.
+     */
+    protected $graphVersion;
+
+    /**
+     * @param FacebookApp $app
+     * @param FacebookClient $client
+     * @param string $accessToken
+     * @param string $graphVersion
+     */
+    public function __construct(FacebookApp $app, FacebookClient $client, $accessToken, $graphVersion)
+    {
+        $this->app = $app;
+        $this->client = $client;
+        $this->accessToken = $accessToken;
+        $this->graphVersion = $graphVersion;
+    }
+
+    /**
+     * Upload by chunks - start phase
+     *
+     * @param string $endpoint
+     * @param FacebookFile $file
+     *
+     * @return FacebookTransferChunk
+     *
+     * @throws FacebookSDKException
+     */
+    public function start($endpoint, FacebookFile $file)
+    {
+        $params = [
+            'upload_phase' => 'start',
+            'file_size' => $file->getSize(),
+        ];
+        $response = $this->sendUploadRequest($endpoint, $params);
+
+        return new FacebookTransferChunk($file, $response['upload_session_id'], $response['video_id'], $response['start_offset'], $response['end_offset']);
+    }
+
+    /**
+     * Upload by chunks - transfer phase
+     *
+     * @param string $endpoint
+     * @param FacebookTransferChunk $chunk
+     * @param boolean $allowToThrow
+     *
+     * @return FacebookTransferChunk
+     *
+     * @throws FacebookResponseException
+     */
+    public function transfer($endpoint, FacebookTransferChunk $chunk, $allowToThrow = false)
+    {
+        $params = [
+            'upload_phase' => 'transfer',
+            'upload_session_id' => $chunk->getUploadSessionId(),
+            'start_offset' => $chunk->getStartOffset(),
+            'video_file_chunk' => $chunk->getPartialFile(),
+        ];
+
+        try {
+            $response = $this->sendUploadRequest($endpoint, $params);
+        } catch (FacebookResponseException $e) {
+            $preException = $e->getPrevious();
+            if ($allowToThrow || !$preException instanceof FacebookResumableUploadException) {
+                throw $e;
+            }
+
+            // Return the same chunk entity so it can be retried.
+            return $chunk;
+        }
+
+        return new FacebookTransferChunk($chunk->getFile(), $chunk->getUploadSessionId(), $chunk->getVideoId(), $response['start_offset'], $response['end_offset']);
+    }
+
+    /**
+     * Attempts to upload a chunk of a file in $retryCountdown tries.
+     *
+     * @param string $endpoint
+     * @param FacebookTransferChunk $chunk
+     * @param int $retryCountdown
+     *
+     * @return FacebookTransferChunk
+     *
+     * @throws FacebookResponseException
+     */
+    public function maxTriesTransfer($endpoint, FacebookTransferChunk $chunk, $retryCountdown)
```

Comment:
I still think that retrying is like a sweet sugar and it shouldn't be part of this service class, it gives it more responsability that just providing function for each phase of a chunk upload.
Having that sweet sugar provided by the `Facebook` master class makes sense, it's an easy way to do for any standard user.
I think it's up to the advanced user that directly use this service to handle retrying an upload.